### PR TITLE
AVX2 Optimization

### PR DESCRIPTION
**Pull request by Dikay900**
Hi there,
first i am aware of the hal changes being currently made so this PR doesn't need to get merged. If needed i will convert them to the new system aswell.

I want to contribute more avx2 code in the future and wanted to get inital feedback about such changes (if there is something special to think of or even if u do not want to get such code contributions if you want to develop it yourself, additional requirements, etc., etc.) before i start converting more code.

please let me know, i am fine with everything.

tests for changed methods are passing on my machine i7-4710MQ (Haswell)

tl;dr; i just want to have feedback if such help is appreciated or not and if this is the correct way; commit doesn't have get merged cause of hal changes

### CODE REVIEW

File: modules/imgproc/src/imgwarp.cpp [hunk 1]
```diff
@@ -4288,10 +4290,32 @@ class RemapInvoker :
...
+                                    ix0 = _mm256_packs_epi32(ix0, ix1);
+                                    iy0 = _mm256_packs_epi32(iy0, iy1);
+                                    ix1 = _mm256_unpacklo_epi16(ix0, iy0);
+                                    iy1 = _mm256_unpackhi_epi16(ix0, iy0);
+                                    _mm256_storeu_si256((__m256i*)(XY + x1 * 2), ix1);
+                                    _mm256_storeu_si256((__m256i*)(XY + x1 * 2 + 16), iy1);
+                                }
+                                _mm256_zeroupper();
+                            }
+                        #elif CV_SSE2
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/imgwarp.cpp]: SSE2 and AVX are compatible. So, you don't need `else`, you can use `if` here too.

Comment (Dikay900) [inline modules/imgproc/src/imgwarp.cpp]: so it will be ok if i do this everywhere where both is used and other amount of work is processed, so that sse code will be used where possible?

Comment (ilya-lavrenov) [inline modules/imgproc/src/imgwarp.cpp]: 1. yes.
2. SSE & AVX are compatible, because they relate to x86/x64. But NEON is ARM specific and not compatible with SSE&AVX, so we use `elif` in this case.

Comment (Dikay900) [inline modules/imgproc/src/imgwarp.cpp]: so we should rather do
```
if neon
else
    if avx
    endif
    if sse
    endif
endif
```

than
```
if neon
endif
if avx
endif
if sse
endif
```
?

Comment (ilya-lavrenov) [inline modules/imgproc/src/imgwarp.cpp]: Actually, the both OK. But I prefer the first one, because in this case we reduce amount of condition compiler need to check when compile for ARM.

Comment (Dikay900) [inline modules/imgproc/src/imgwarp.cpp]: ok then i will do that thanks for your feedback

**Comment by vpisarev** [comment]
@Dikay900, we need to weight all pro's and con's regarding AVX2.
1) at the moment we do not have any AVX2-capable hardware in our buildbot, therefore we can not test this code.
2) Intel gives us and all our users a subset of IPP for free (at no charge, including zero license fee for opencv-based applications), where they use all those fancy instructions. I'd let them do that.
3) we could have add AVX2 backend for our universal intrinsics (opencv2/hal/intrin_avx2.hpp), but the problem is that the vector size is 256 bits, not 128 and the other problem is that this AVX code still has to be dispatched at runtime. With SSE2 (even SSSE3) and NEON we can safely assume that 99% of CPUs support it. Exotic cases like Raspberry Pi 1 or Intel Quark would need special builds of OpenCV anyway. Therefore, the universal intrinsics can be used without any #if ... and that was one of the main goal when designing them. With AVX we need to wait 5-10 more years until it's universally available. So we still have complete duplication of branches, which needs to be tested.

Therefore, I would say, my current position is that we should not include AVX2 code on general basis. However, if for certain "heavy" functions where performance measurements are done and AVX2 demonstrates dramatic speedup over SSE2/SSSE3 and over IPP (say, 20% or more), such optimizations can be included.

**Comment by Dikay900** [comment]
If i find something of a huge speed up i will try to contribute this, but think 20% won't be possible in most cases.

Thank you for this message, closing this then.

**Comment by ilya-lavrenov** [comment]
@vpisarev, I see that you are skeptical regarding AVX2, but I support @Dikay900, and I believe that he does a good job. So, my suggestions are:
* perform all AVX/AVX2 optimizations in separate compilation units compiled with AVX/AVX2 compilation options. For example, we have `avx2_canny` in `canny_avx2.cpp`.
* from usual source files we will use checkHardwareSupport(CV_CPU_AVX2) and invoke AVX2-optimized functions only if the check is passed. For example:
```
// canny.cpp
if (checkHardwareSupport(CV_CPU_AVX2))
{
    avx2_canny(..);
    return;
} 
..
```
And binaries compiled in such manner will work on hardware without AVX2 support (actually, in 2.4 branch we have the same scheme for image warping operations).

But as you said our buildbot infrastructure doesn't have Haswell or better, so, we can disable AVX2 by default and all public packages will be built without AVX2.
And @Dikay900 can continue his optimizations and at the time when AVX2 will universally available we will have a set of AVX2 optimized functions, that not covered by IPP. It's better than nothing.

**Comment by Dikay900** [comment]
i think the main point of @vpisarev was that the universal intrinsics cannot get implemented safely.
On the other hand i think we do not explicitly need this universal instrinsics right now since we can develop every SSE code to AVX code or am i missing something with these universal intrinsics?

Since we've got additional ifs everywhere where sse or avx code is available the code will still run on any hardware so i think there is no need for additional compilation units, which would create much additional redundant code.

The problem with general AVX2 code remains that there is no automatic testing environment. Another build configuration for AVX code is needed there to cover all AVX and SSE code (since AVX can eliminate SSE code in some cases completly)

On the other hand users may want the most out of their cpus for very cpu intensive tasks.
Maybe we can add an additional AVX2_experimental option into the definiton which can be used for this kind of code defaulting to 0 which will use the normal SSE code.

I think the biggest problem here is that the test enviroment cannot test this new code and i think this is the main problem here, for everything else there would be solution.

Thanks for your support @ilya-lavrenov

**Comment by ilya-lavrenov** [comment]
> I think the main point of @vpisarev was that the universal intrinsics cannot get implemented safely.

My opinion is that everybody is free to write the code using SSE/AVX/NEON intrstions etc ***explicitly***, not using universal intrinsics. They are not tested ***at all*** and in experimental phase. So, we must not reject any explicit optimizations!

> Since we've got additional ifs everywhere where sse or avx code is available the code will still run on any hardware so i think there is no need for additional compilation units, which would create much additional redundant code.

No, simple `if` in the code is not enough. When you compile all the sources with AVX2 option, compiler automatically can use AVX2 instructions without your knowledge and this binary file will give `abort trap: illegal instruction`. So, a separate compilation units is necessary to distinct AVX code and another one.

@vpisarev, we are still waiting your voice. I believe that we can implement such infrastructure inside a HAL.

**Comment by Dikay900** [comment]
>My opinion is that everybody is free to write the code using SSE/AVX/NEON intrstions etc explicitly, not using universal intrinsics.
yes thats also my opinion, could be that my statements to wait with this code for the hal implementation was a bit confuing since i did not know what there would have been into the hal module.
> No, simple if in the code is not enough.
I thought the checkHardwareSupport method will make a realtime estimation what is supported?
If this is wrong why is this used anyway since we have macros and such methods for this.

**Comment by ilya-lavrenov** [comment]
> I thought the checkHardwareSupport method will make a realtime estimation what is supported?

It will work if we run SSE4 compiled code on SSE2 compatible device. Bit it doesn't work if we run AVX code on SSE-compatible device. Because AVX it's a "next generation" (256-bit vectors are not compatible with 128-bits ones). And we need additional mechanism like compilation units.

**Comment by Dikay900** [comment]
i understood already that we need avx supported hardware for specific features but in this code i added we have the following
```
#if CV_AVX2 // AVX2 enabled via Cmake / Compiler
    bool useAVX2 = checkHardwareSupport(CV_CPU_AVX2); // realtime estimation if hardware supports AVX2
#endif
...
#if CV_AVX2
     if (useAVX2) // will only run if AVX2 is enabled and supported by hardware
```

am i wrong with my comments in this snippet? I thought with this mechanism it is safe to assume that code will only run on supported hardware (thus you need AVX2 capable cpu to compile it)

**Comment by vpisarev** [comment]
@Dikay900,
universal intrinsics are aimed at providing portable and reasonably efficient wrappers around different SIMD engines. That is, instead of
#if CV_SSE2
     for(...)
         _mm_storeu_ps(c + i, _mm_add_ps(_mm_loadu_ps(a + i), _mm_loadu_ps(b + i));
#elif CV_NEON
     for(...)
     vst1q_f32(c + i, vaddq_f32(vld1q_f32(a + i), vld1q_f32(b + i));
#else ...
     for( int i = 0; i < n; i++ ) c[i] = a[i] + b[i];
#endif
you simply write (in the current version)
#if CV_SIMD128
     for(...) v_store(c + i, v_load(a+i) + v_load(b+i));
#else
     for( int i = 0; i < n; i++ ) c[i] = a[i] + b[i];
#endif

and even this #if is not strictly needed, because the intrinsics are also emulated in C.
I actually would like to get rid of, or at least reduce the number of #if's in OpenCV code, because they really make the thorough testing more difficult. This is possible with SSE (SSSE3 to be precise) and NEON, since they are available universally. AVX does not fit into this scheme, unfortunately. Here I'm talking about all opencv and opencv_contrib modules with 1 notable exception - the hal module itself.

As you may note, we also have opencv2/hal.hpp with a banch of cv::hal::* functions. We hope to increase this number dramatically during the next several months. And that's where we can use the idea suggested by Ilya. Namely, certain cv::hal functions may include SSE4, AVX2 etc. branches and do runtime dispatching there; to play safe, AVX2-based code can be put into separate files and built with special options not to let compiler use AVX2 instructions unconditionally.

So the idea is to hide #if's in the main OpenCV inside universal intrinsics. Performance critical things can be put into HAL and optimized using fastest CPU instructions available.

@ilya-lavrenov,
regarding the comment that everyone is free to use intrinsics directly. Well, I'd tighten these rules because of the obvious reason - "long-term support". We happily use the universal cv::parallel_for_ instead of a wild combination of openmp, TBB etc. I do not see any reasons why we should do the same with universal intrinsics. I looked at the generated code and at least with SSE modern compilers do an excellent job with turning these C++ intrinsics into efficient assembly instructions, as if you used assembly language directly. So, I think, we should encourage people to use universal intrinsics for most of not all the newly written code.

**Comment by Dikay900** [comment]
ok understood. Thank you very much for the further explanation.

**Comment by vpisarev** [comment]
@Dikay900, thank you too. For now I'm closing this PR and feel free to reopen it when we have more solid HAL and develop methodology of putting AVX and other specialized optimizations into separate files for safe runtime dispatching.
